pkg/filter/bloom_filter.go

```diff
@@ -0,0 +1,145 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Package filter defines and implements data structures and interfaces for skipping index.
+package filter
+
+import (
+	"errors"
+	"hash"
+	"hash/fnv"
+	"math"
+
+	"github.com/bits-and-blooms/bitset"
+)
+
+// FalsePositiveRate is the default false positive rate for Bloom filter.
+const FalsePositiveRate = 0.01
+
+// BloomFilter is a probabilistic data structure designed to test whether an element is a member of a set.
+type BloomFilter struct {
+	hashFunc hash.Hash64
+	bits     *bitset.BitSet
+	m        uint64
+	k        uint
+}
+
+// NewBloomFilter creates a new Bloom filter with the number of items n and false positive rate p.
+func NewBloomFilter(n uint64, p float64) (*BloomFilter, error) {
+	if p <= 0 || p >= 1 {
+		return nil, errors.New("invalid false positive rate")
+	}
+	if n <= 0 {
+		return nil, errors.New("invalid number of items")
+	}
+	m := calculateM(n, p)
+	k := calculateK(n, m)
```

Comment:
The modern bloom filter typically uses fixed values for bits per item (b) and the number of hash functions (k) to derive the probability of a false positive (p). The relationship can be expressed with the following formula:

```
p ≈ (1 - e^(-k/b))^k
```

If we want the probability (p) to equal 1%, suitable values for b and k would be 10 and 7, respectively. Consequently, the total number of bits (m) can be calculated as m = b* n . This approach allows us to avoid the CPU overhead associated with floating-point calculations.